### PR TITLE
builtins: add `Core.throw_methoderror`

### PR DESCRIPTION
**Pull request by topolarity**
This allows us to simulate/mark calls that are known-to-fail.

Required for https://github.com/JuliaLang/julia/pull/54972/

### CODE REVIEW

File: base/compiler/abstractinterpretation.jl [hunk 1]
```diff
@@ -2226,6 +2226,18 @@ function abstract_throw(interp::AbstractInterpreter, argtypes::Vector{Any}, ::Ab
...
+function abstract_throw_methoderror(interp::AbstractInterpreter, argtypes::Vector{Any}, ::AbsIntState)
+    na = length(argtypes)
+    exct = if na == 1
+        ArgumentError
+    elseif !isvarargtype(argtypes[2])
+        MethodError
+    else
+        tmerge(𝕃ᵢ, MethodError, ArgumentError)
+    end
+    return CallMeta(Union{}, exct, EFFECTS_THROWS, NoCallInfo())
```
Comment (gbaraldi) [inline base/compiler/abstractinterpretation.jl:2238]: I guess its not consistent because it captures the world?

Comment (topolarity) [reply base/compiler/abstractinterpretation.jl:2238]: Great point - I got that right in the tfuncs but not here.

Comment (topolarity) [reply base/compiler/abstractinterpretation.jl]: Actually, I think this is consistent after all.

Consistency doesn't place requirements on exceptions:
```
   If :consistent functions terminate by throwing an exception, that exception
    itself is not required to meet the egality requirement specified above.
```

And even if it did, it is per-world-age:
```
 The :consistent-cy assertion is made world-age wise. More formally, write
    fᵢ for the evaluation of f in world-age i, then this setting requires:
    ∀ i, x, y: x ≡ y → fᵢ(x) ≡ fᵢ(y)
    However, for two world ages i, j s.t. i ≠ j, we may have fᵢ(x) ≢ fⱼ(y).
```

Comment (aviatesk) [reply base/compiler/abstractinterpretation.jl:2237]: Yeah, `:consistent` doesn't require the consistency on thrown object. If a method throws consistently, then it is `:consistent` no matter what it throws.

File: src/builtins.c [hunk 2]
```diff
@@ -580,6 +580,14 @@ JL_CALLABLE(jl_f_throw)
     return jl_nothing;
 }
 
+JL_CALLABLE(jl_f_throw_methoderror)
+{
+    JL_NARGSV(throw_methoderror, 1);
+    size_t world = jl_get_tls_world_age();
+    jl_method_error((jl_function_t *)args[0], &args[1], nargs, world);
```
Comment (aviatesk) [inline src/builtins.c:587]:
```suggestion
    jl_method_error((jl_value_t *)args[0], &args[1], nargs, world);
```
since arbitrary object can be a functor?

**Review by aviatesk — APPROVED** [review]

File: base/compiler/abstractinterpretation.jl (hunk 1 above)
Comment (oscardssmith) [reply base/compiler/abstractinterpretation.jl:2237]: wait that is a kind of odd spec for it to have. Does that mean we can't inline the throw of an exception if we think a function throws?

Comment (aviatesk) [reply base/compiler/abstractinterpretation.jl:2237]: Yeah, even if concrete evaluation reveals that a method call will definitely throw, we can't propagate that exception as a `Const`.

**Review by aviatesk — APPROVED** [review]

**Pull request merged**